pkg/api/agentPoolOnlyApi/v20180331/validate.go

```diff
@@ -141,108 +141,121 @@ func validateVNET(a *Properties) error {
 
 	n := a.NetworkProfile
 
-	if n == nil {
-		return ErrorNilNetworkProfile
-	}
+	// validate network profile settings
+	if n != nil {
+		if string(n.NetworkPlugin) == string(Azure) || string(n.NetworkPlugin) == string(Kubenet) {
+			if n.ServiceCidr != "" && n.DNSServiceIP != "" && n.DockerBridgeCidr != "" {
+				// validate ServiceCidr
+				_, serviceCidr, err := net.ParseCIDR(n.ServiceCidr)
+				if err != nil {
+					return ErrorInvalidServiceCidr
+				}
 
-	if string(n.NetworkPlugin) == string(Azure) {
-		// validate ServiceCidr
-		_, serviceCidr, err := net.ParseCIDR(n.ServiceCidr)
-		if err != nil {
-			return ErrorInvalidServiceCidr
-		}
+				// validate DNSServiceIP
+				dnsServiceIP := net.ParseIP(n.DNSServiceIP)
+				if dnsServiceIP == nil {
+					return ErrorInvalidDNSServiceIP
+				}
 
-		// validate DNSServiceIP
-		dnsServiceIP := net.ParseIP(n.DNSServiceIP)
-		if dnsServiceIP == nil {
-			return ErrorInvalidDNSServiceIP
-		}
+				// validate DockerBridgeCidr
+				_, _, err = net.ParseCIDR(n.DockerBridgeCidr)
+				if err != nil {
+					return ErrorInvalidDockerBridgeCidr
+				}
 
-		// validate DockerBridgeCidr
-		_, _, err = net.ParseCIDR(n.DockerBridgeCidr)
-		if err != nil {
-			return ErrorInvalidDockerBridgeCidr
-		}
+				// validate DNSServiceIP is within ServiceCidr
+				if !serviceCidr.Contains(dnsServiceIP) {
+					return ErrorDNSServiceIPNotInServiceCidr
+				}
 
-		// validate DNSServiceIP is within ServiceCidr
-		if !serviceCidr.Contains(dnsServiceIP) {
-			return ErrorDNSServiceIPNotInServiceCidr
+				// validate DNSServiceIP is not the first IP in ServiceCidr. The first IP is reserved for redirect svc.
+				kubernetesServiceIP, err := common.CidrStringFirstIP(n.ServiceCidr)
+				if err != nil {
+					return ErrorInvalidServiceCidr
+				}
+				if dnsServiceIP.String() == kubernetesServiceIP.String() {
+					return ErrorDNSServiceIPAlreadyUsed
+				}
+			} else if n.ServiceCidr == "" && n.DNSServiceIP == "" && n.DockerBridgeCidr == "" {
+				// this is a valid case, and no validation needed.
+			} else {
+				return ErrorInvalidNetworkProfile
+			}
+		} else {
+			return ErrorInvalidNetworkPlugin
 		}
+	}
 
-		// validate DNSServiceIP is not the first IP in ServiceCidr. The first IP is reserved for redirect svc.
-		kubernetesServiceIP, err := common.CidrStringFirstIP(n.ServiceCidr)
-		if err != nil {
-			return ErrorInvalidServiceCidr
-		}
-		if dnsServiceIP.String() == kubernetesServiceIP.String() {
-			return ErrorDNSServiceIPAlreadyUsed
+	// validate agent pool custom VNET settings
+	if a.AgentPoolProfiles != nil {
+		if e := validateAgentPoolVNET(a.AgentPoolProfiles); e != nil {
+			return e
 		}
+	} else {
+		return ErrorNilAgentPoolProfile
```

Comment:
In Update case, it may not be required.

Reply:
Yes, it is probably redundant and has been checked before, but I think it is no harm to be on the safe side to check this.

Reply:
if you are sure that update without `AgentPoolProfile` will be fine, then it's ok...

Reply:
cant this be done at https://github.com/JunSun17/acs-engine/blob/65d35f4907534ef3db028fdd4302c03f62bbf17e/pkg/api/agentPoolOnlyApi/v20180331/validate.go#L20 ?

Reply:
Since we allow partial API model in the update case, remove ErrorNilAgentPoolProfile.

I still want to leave the validateAgentPoolVNET() method in VNET validation since VNET validation will cross multiple profiles (for example, we will need to validate subnetCidr from agentPoolProfile against serviceCidr from networkProfile).